demos/digest/EVP_f_md.c

```diff
@@ -0,0 +1,121 @@
+/*-
+ * Copyright 2019-2021 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/*
+ * Example of using EVP_MD_fetch and EVP_Digest* methods to calculate
+ * a digest of static buffers
+ * You can find SHA3 test vectors from NIST here:
+ * https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Algorithm-Validation-Program/documents/sha3/sha-3bytetestvectors.zip
+ * Use xxd convert a hex input:
+ * echo "1ca984dcc913344370cf" | xxd -r -p | ./EVP_MD_stdin
+ */
+
+#include <string.h>
+#include <stdio.h>
+#include <openssl/err.h>
+#include <openssl/bio.h>
+#include <openssl/evp.h>
+
+/*-
+ * This demonstration will show how to digest data using
+ * a BIO configured with a message digest
+ * A message digest name may be passed as an argument.
+ * The default digest is sha3-512
+ */
+
+int main(int argc, char * argv[])
+{
+    int result = 1;
+    OSSL_LIB_CTX *library_context = NULL;
+    BIO *input = NULL;
+    BIO *bio_digest = NULL;
+    const char *digest_name;
+    EVP_MD *md = NULL;
+    unsigned char buffer[512];
+    size_t readct, writect;
+    unsigned char *digest_value=NULL;
+    int j;
+
+    if (argc > 1)
+        digest_name = argv[1];
+    else 
+        digest_name = "sha3-512";
+    
+    input = BIO_new_fd( fileno(stdin), 1 );
```

Comment:
It could use a memory bio here instead using fixed data?

Reply:
Not sure I understand. The app is reading from stdin.

Reply:
So the question was - should it read from stdin or should we just have a fixed buffer that we read in chunks from?
i.e are we making an interactive application or should it be self contained?

Reply:
My intention was in the comment
/*=
 * Example of using EVP_MD_fetch and EVP_Digest* methods to calculate
 * a digest of static buffers
 * You can find SHA3 test vectors from NIST here:
 * https://csrc.nist.gov/CSRC/media/Projects/Cryptographic-Algorithm-Validation-Program/documents/sha3/sha-3bytetestvectors.zip
 * For example, contains these lines:
    Len = 80
    Msg = 1ca984dcc913344370cf
    MD = 6915ea0eeffb99b9b246a0e34daf3947852684c3d618260119a22835659e4f23d4eb66a15d0affb8e93771578f5e8f25b7a5f2a55f511fb8b96325ba2cd14816
 * use xxd convert the hex message string to binary input for EVP_MD_stdin:
 * echo "1ca984dcc913344370cf" | xxd -r -p | ./EVP_f_md
 * and then verify the output matches MD above.
 */

Reply:
I am not overly keen on making the demos interactive applications with lots of external inputs.
See the KISS principle :)
Maybe this is a short OTC discussion?

Reply:
+1 to @slontism suggestion. The demos should be as simple as possible. There should be no distracting code around.